src/Messages/TSBlockingManager.m

```diff
@@ -0,0 +1,239 @@
+//
+
+#import "TSBlockingManager.h"
+#import "OWSBlockedPhoneNumbersMessage.h"
+#import "OWSMessageSender.h"
+#import "TSStorageManager.h"
+#import "TextSecureKitEnv.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+NSString * const kNSNotificationName_BlockedPhoneNumbersDidChange = @"kNSNotificationName_BlockedPhoneNumbersDidChange";
+NSString * const kTSStorageManager_BlockedPhoneNumbersCollection = @"kTSStorageManager_BlockedPhoneNumbersCollection";
+// This key is used to persist the current "blocked phone numbers" state.
+NSString * const kTSStorageManager_BlockedPhoneNumbersKey = @"kTSStorageManager_BlockedPhoneNumbersKey";
+// This key is used to persist the most recently synced "blocked phone numbers" state.
+NSString *const kTSStorageManager_SyncedBlockedPhoneNumbersKey = @"kTSStorageManager_SyncedBlockedPhoneNumbersKey";
+
+@interface TSBlockingManager ()
+
+@property (nonatomic, readonly) TSStorageManager *storageManager;
+@property (nonatomic, readonly) OWSMessageSender *messageSender;
+
+// We don't store the phone numbers as instances of PhoneNumber to avoid
+// consistency issues between clients, but these should all be valid e164
+// phone numbers.
+@property (nonatomic, readonly) NSMutableSet<NSString *> *blockedPhoneNumberSet;
+
+@end
+
+#pragma mark -
+
+@implementation TSBlockingManager
+
++ (instancetype)sharedManager {
+    static TSBlockingManager *sharedMyManager = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedMyManager = [[self alloc] initDefault];
+    });
+    return sharedMyManager;
+}
+
+- (instancetype)initDefault
+{
+    TSStorageManager *storageManager = [TSStorageManager sharedManager];
+    OWSMessageSender *messageSender = [TextSecureKitEnv sharedEnv].messageSender;
+
+    return [self initStorageManager:storageManager
+                      messageSender:messageSender];
+}
+
+- (instancetype)initStorageManager:(TSStorageManager *)storageManager messageSender:(OWSMessageSender *)messageSender
+{
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    OWSAssert(storageManager);
+    OWSAssert(messageSender);
+    
+    _storageManager = storageManager;
+    _messageSender = messageSender;
+
+    [self loadBlockedPhoneNumbers];
+
+    return self;
+}
+
+
+- (void)addBlockedPhoneNumber:(NSString *)phoneNumber {
+    OWSAssert(phoneNumber.length > 0);
+
+    DDLogInfo(@"%@ addBlockedPhoneNumber: %@", self.tag, phoneNumber);
+
+    @synchronized (self) {
+        if ([_blockedPhoneNumberSet containsObject:phoneNumber]) {
+            return;
+        }
+        
+        [_blockedPhoneNumberSet addObject:phoneNumber];
+    }
+
+    [self handleUpdate:NO];
+}
+
+- (void)removeBlockedPhoneNumber:(NSString *)phoneNumber {
+    OWSAssert(phoneNumber.length > 0);
+
+    DDLogInfo(@"%@ removeBlockedPhoneNumber: %@", self.tag, phoneNumber);
+
+    @synchronized (self) {
+        if (![_blockedPhoneNumberSet containsObject:phoneNumber]) {
```

Comment:
nit: At first I wondered why you bothered with this "redundant" set logic, but I realized the goal of this early return is to avoid the call to `handleUpdate`. Maybe worth a comment here an in `addBlockedPhoneNumber`